crates/bevy_ecs/src/lib.rs

```diff
@@ -84,6 +89,17 @@ pub mod prelude {
     pub use crate::reflect::AppFunctionRegistry;
 }
 
+/// Exports used by macros.
+///
+/// These are not meant to be used directly and are subject to breaking changes.
+#[doc(hidden)]
+pub mod __macro_exports {
+    // Cannot directly use `alloc::vec::Vec` in macros, as a crate may not have
+    // included `extern crate alloc;`. This re-export ensures we have access
+    // to `Vec` in `no_std` and `std` contexts.
+    pub use alloc::vec::Vec;
+}
```

Comment:
The new required components recursion check uses a `Vec`, which isn't in the global namespace with `no_std`. This technique of adding a `pub mod __macro_exports { ... }` is how `bevy_reflect` gets around this issue, so this is at least consistent, even if it isn't ideal.

Reply:
We could define a feature-flagged type alias here instead right? But yeah, consistent is fine for now.